src/main/java/com/microsoft/sqlserver/jdbc/SQLServerAASEnclaveProvider.java

```diff
@@ -0,0 +1,493 @@
+                            null, 0, e);
+                }
+            }
+
+            // Process the third resultset.
+            if (connection.isAEv2() && stmt.getMoreResults()) {
+                rs = (SQLServerResultSet) stmt.getResultSet();
+                while (rs.next()) {
+                    hgsResponse = new AASAttestationResponse(rs.getBytes(1));
+                    // This validates and establishes the enclave session if valid
+                    if (!connection.enclaveEstablished()) {
+                        hgsResponse = validateAttestationResponse(hgsResponse);
+                    }
+                }
+            }
+
+            // Null check for rs is done already.
+            rs.close();
+        } catch (SQLException e) {
+            if (e instanceof SQLServerException) {
+                throw (SQLServerException) e;
+            } else {
+                throw new SQLServerException(SQLServerException.getErrString("R_UnableRetrieveParameterMetadata"), null,
+                        0, e);
+            }
+        }
+        return enclaveRequestedCEKs;
+    }
+
+    @Override
+    public void setEnclaveSession(EnclaveCacheEntry entry) {
+        this.enclaveSession = entry.getEnclaveSession();
+        this.aasParams = (AASAttestationParameters) entry.getBaseAttestationRequest();
+    }
+}
+
+
+class AASAttestationParameters extends BaseAttestationRequest {
+
+    // Type 1 is AAS, sent as Little Endian 0x10000000
+    private static byte ENCLAVE_TYPE[] = new byte[] {0x1, 0x0, 0x0, 0x0};
+    // Nonce length is always 256
+    private static byte NONCE_LENGTH[] = new byte[] {0x0, 0x1, 0x0, 0x0};
+    private byte[] nonce = new byte[256];
+
+    AASAttestationParameters(String attestationUrl) throws SQLServerException, IOException {
+        byte[] attestationUrlBytes = (attestationUrl + '\0').getBytes(UTF_16LE);
+
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        os.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(attestationUrlBytes.length).array());
+        os.writeBytes(attestationUrlBytes);
+        os.writeBytes(NONCE_LENGTH);
+        new SecureRandom().nextBytes(nonce);
+        os.writeBytes(nonce);
+        enclaveChallenge = os.toByteArray();
+
+        initBcryptECDH();
+    }
+
+    @Override
+    byte[] getBytes() {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        os.writeBytes(ENCLAVE_TYPE);
+        os.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(enclaveChallenge.length).array());
+        os.writeBytes(enclaveChallenge);
+        os.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(ENCLAVE_LENGTH).array());
+        os.writeBytes(ECDH_MAGIC);
+        os.writeBytes(x);
+        os.writeBytes(y);
+        return os.toByteArray();
+    }
+
+    byte[] getNonce() {
+        return nonce;
+    }
+}
+
+
+class JWTCertificateEntry {
```

Comment:
general comment (applies elsewhere too)
Can we add comments to these classes? I know they're not public so no need for javadocs, but would be nice to have some descriptions on what these are